File: application/app/domain/model/Endpoint.scala
package domain.model

import common.validation.Validatable
import domain.exceptions.RouteNotFoundException
import domain.model.AuthenticationMode.AuthenticationMode
import domain.model.RouteType.RouteType
import javax.validation.Valid
import javax.validation.constraints.{NotBlank, NotNull, Pattern}

import scala.annotation.meta.field

case class Endpoint(
                     id: String,
                     @(NotBlank @field) @(Pattern @field)(regexp="^(?!\\/api\\/internal).*") path: String,
                     @(Valid @field) @(NotNull @field) private val configuration: EndpointConfiguration
                   ) extends Validatable {
  def getConnectRoute: Route = getRoute(RouteType.CONNECT) match {
    case Some(r) => r
    case None => throw RouteNotFoundException("Connect route is not defined")
  }

  def getDisconnectRoute: Route = getRoute(RouteType.DISCONNECT) match {
    case Some(r) => r
    case None => throw RouteNotFoundException("Disconnect route is not defined")
  }

  def getDefaultRoute: Route = getRoute(RouteType.DEFAULT) match {
    case Some(r) => r
    case None => throw RouteNotFoundException("Default route is not defined")
  }

  def getCustomRoutes: Set[Route] = getRoutes(RouteType.CUSTOM)

  def findCustomRoute(json: String): Option[Route] = getCustomRoutes.find(r => r.appliesTo(json))

  def getRoute(routeType: RouteType): Option[Route] = configuration.routes.find(_.routeType == routeType)
  def getRoutes(routeType: RouteType): Set[Route] = configuration.getRoutes(routeType)
  def hasRoute(routeType: RouteType): Boolean = configuration.routes.exists(_.routeType == routeType)

  def addRoutes(routes: Set[Route]): Endpoint = {
    val newConfig = configuration.copy(routes = configuration.routes.++(routes))
    copy(configuration = newConfig)
  }

  def routes: Set[Route] = configuration.routes
  def filters: Set[Filter] = configuration.filters
  def authenticationMode: AuthenticationMode = configuration.authenticationMode

  def bufferSize: Int = configuration.bufferSize
  def backendParallelism: Int = configuration.backendParallelism

  def matchesPath(targetPath: String): Boolean = targetPath.matches(path)
}

object Endpoint {
  def apply(id: String, path: String): Endpoint = {
    new Endpoint(id, path, EndpointConfiguration(Set.empty, Set.empty))
  }

  def apply(id: String, path: String, endpointConfiguration: EndpointConfiguration): Endpoint = {
    new Endpoint(id, path, endpointConfiguration)
  }

  def apply(id: String, path: String, filters: Set[Filter], routes: Set[Route]): Endpoint = {
    new Endpoint(id, path, EndpointConfiguration(filters, routes))
  }

  def apply(id: String, path: String, filters: Set[Filter], routes: Set[Route], authenticationMode: AuthenticationMode): Endpoint = {
    new Endpoint(id, path, EndpointConfiguration(filters, routes, authenticationMode))
  }
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/Endpoint.java
package com.cosmin.wsgateway.domain;

import static com.cosmin.wsgateway.domain.Route.Type.CONNECT;
import static com.cosmin.wsgateway.domain.Route.Type.CUSTOM;
import static com.cosmin.wsgateway.domain.Route.Type.DEFAULT;
import static com.cosmin.wsgateway.domain.Route.Type.DISCONNECT;

import com.cosmin.wsgateway.domain.exceptions.RouteNotFoundException;
import java.util.LinkedHashSet;
import java.util.Optional;
import java.util.Set;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;
import lombok.Builder;
import lombok.Value;
import lombok.With;

@Value
@Builder
@With
public class Endpoint {
    private final String id;

    @NotNull
    @Size(min = 2, max = 255)
    @Pattern(regexp = "^(?!\\/api\\/internal).*")
    private final String path;

    @NotNull
    @Valid
    private final EndpointConfiguration configuration;

    public Optional<Route> findCustomRoute(String json) {
        return configuration.getRoutesByType(CUSTOM)
                .stream()
                .filter(r -> r.appliesTo(json))
                .findFirst();
    }

    public Route getConnectRoute() {
        return getRoute(CONNECT).orElseThrow(() -> new RouteNotFoundException("Connect Route is not defined"));
    }

    public Route getDisconnectRoute() {
        return getRoute(DISCONNECT).orElseThrow(() -> new RouteNotFoundException("Disconnect Route is not defined"));
    }

    public Route getDefaultRoute() {
        return getRoute(DEFAULT).orElseThrow(() -> new RouteNotFoundException("Default Route is not defined"));
    }

    public Boolean hasRoute(Route.Type type) {
        return getRoute(type).isPresent();
    }

    public Optional<Route> getRoute(Route.Type type) {
        return configuration.getRoutesByType(type)
                .stream()
                .filter(r -> r.getType().equals(type))
                .findFirst();
    }

    public Endpoint addRoutes(Set<Route> routes) {
        Set<Route> newRoutes = new LinkedHashSet<>(getRoutes());
        newRoutes.addAll(routes);
        EndpointConfiguration newConfiguration = configuration.withRoutes(newRoutes);

        return withConfiguration(newConfiguration);
    }

    public Set<Route> getRoutes() {
        return configuration.getRoutes();
    }

    public Authentication getAuthentication() {
        return configuration.getAuthentication();
    }

    public GeneralSettings getGeneralSettings() {
        return configuration.getGeneralSettings();
    }

    public Set<Filter<?>> getFilters() {
        return configuration.getFilters();
    }

    public Boolean matchesPath(String targetPath) {
        return targetPath.matches(path);
    }
}
